File: frontend/src/lib/api/platform-access.ts
import { z } from 'zod'

import {
  AnalysisApiError,
  analysisRequest,
  decodeAnalysis,
  jsonMutation,
  safeId,
} from '@/lib/api/analysis-shared'
import { isoDateSchema } from '@/lib/api/search-runs'

export const PLATFORM_ACCESS_QUERY_KEY = ['platform-access-pacing'] as const
export const PLATFORM_ACCESS_PLATFORMS = [
  'wb',
  'dy',
  'ks',
  'xhs',
  'toutiao',
] as const

export const platformIntervalsSchema = z.strictObject({
  wb: z.number().int().min(1).max(300),
  dy: z.number().int().min(1).max(300),
  ks: z.number().int().min(1).max(300),
  xhs: z.number().int().min(1).max(300),
  toutiao: z.number().int().min(1).max(300),
})
export const platformAccessSnapshotSchema = z.strictObject({
  interval_seconds: platformIntervalsSchema,
  basis: z.enum(['explicit', 'upgrade_safe_default', 'legacy_unavailable']),
})
export const platformAccessDiagnosticSchema = z.strictObject({
  platform: z.enum(PLATFORM_ACCESS_PLATFORMS),
  stage: z.enum(['search', 'detail', 'media']),
  outcome: z.literal('platform_blocked_or_rate_limited'),
  basis: z.enum([
    'http_status',
    'explicit_platform_evidence',
    'browser_dom_evidence',
  ]),
  status_code: z.number().int().min(100).max(599).nullable(),
  platform_code: z.string().min(1).max(100).nullable(),
  retry_after_at: z.string().datetime({ offset: true }).nullable(),
  manual_challenge_required: z.boolean(),
  observed_at: z.string().datetime({ offset: true }),
})
const platformAccessSettingsSchema = z.strictObject({
  revision: safeId,
  interval_seconds: platformIntervalsSchema,
  updated_at: isoDateSchema,
})
export type PlatformAccessIntervals = z.infer<typeof platformIntervalsSchema>
export type PlatformAccessSnapshot = z.infer<
  typeof platformAccessSnapshotSchema
>
export type PlatformAccessSettings = z.infer<
  typeof platformAccessSettingsSchema
>
export type PlatformAccessUpdate = {
  expected_revision: number
  interval_seconds: PlatformAccessIntervals
}

export async function fetchPlatformAccessSettings(signal: AbortSignal) {
  return decodeAnalysis(
    platformAccessSettingsSchema,
    await analysisRequest('/platform-access-pacing', { signal }),
  )
}

export async function savePlatformAccessSettings(value: PlatformAccessUpdate) {
  const valid = platformAccessSettingsSchema
    .pick({ revision: true })
    .safeParse({ revision: value.expected_revision })
  if (
    !valid.success ||
    !platformIntervalsSchema.safeParse(value.interval_seconds).success
  )
    throw new AnalysisApiError('invalid_request')
  return decodeAnalysis(
    platformAccessSettingsSchema,
    await analysisRequest(
      '/platform-access-pacing',
      jsonMutation('PUT', value),
    ),
  )
}

export function platformAccessErrorMessage(error: unknown) {
  if (error instanceof AnalysisApiError) {
    if (error.code === 'platform_access_settings_conflict')
      return '平台访问间隔已被其他窗口更新，请刷新后重试。'
    if (error.code === 'platform_access_settings_unavailable')
      return '平台访问间隔暂时无法读取或保存，请稍后重试。'
  }
  return '平台访问间隔暂时无法读取或保存，请稍后重试。'
}

File: backend/src/opinion_workbench_api/schemas/platform_access.py
"""Strict local settings and immutable snapshots for platform access pacing."""

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from opinion_workbench_api.search_platforms import SEARCH_PLATFORMS, SearchPlatform

PLATFORM_ACCESS_DEFAULT_SECONDS = 5
PLATFORM_ACCESS_MIN_SECONDS = 1
PLATFORM_ACCESS_MAX_SECONDS = 300
MAX_SAFE_INTEGER = 2**53 - 1
PlatformAccessBasis = Literal["explicit", "upgrade_safe_default", "legacy_unavailable"]


class StrictPlatformModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)


class PlatformIntervals(StrictPlatformModel):
    """A complete, ordered five-platform interval map."""

    wb: int = Field(ge=PLATFORM_ACCESS_MIN_SECONDS, le=PLATFORM_ACCESS_MAX_SECONDS)
    dy: int = Field(ge=PLATFORM_ACCESS_MIN_SECONDS, le=PLATFORM_ACCESS_MAX_SECONDS)
    ks: int = Field(ge=PLATFORM_ACCESS_MIN_SECONDS, le=PLATFORM_ACCESS_MAX_SECONDS)
    xhs: int = Field(ge=PLATFORM_ACCESS_MIN_SECONDS, le=PLATFORM_ACCESS_MAX_SECONDS)
    toutiao: int = Field(ge=PLATFORM_ACCESS_MIN_SECONDS, le=PLATFORM_ACCESS_MAX_SECONDS)

    @model_validator(mode="after")
    def reject_bool_values(self):
        if any(type(value) is not int for value in self.model_dump().values()):
            raise ValueError("platform intervals must be integers")
        return self

    def for_platform(self, platform: SearchPlatform | str) -> int:
        if platform not in SEARCH_PLATFORMS:
            raise ValueError("unsupported platform")
        return int(getattr(self, platform))


class PlatformAccessSnapshot(StrictPlatformModel):
    """The pacing policy captured when a task is admitted."""

    interval_seconds: PlatformIntervals
    basis: PlatformAccessBasis

    @property
    def intervals_seconds(self) -> PlatformIntervals:
        """Compatibility spelling used by a few older internal callers."""
        return self.interval_seconds

    def for_platform(self, platform: SearchPlatform | str) -> int:
        return self.interval_seconds.for_platform(platform)


class PlatformAccessSettings(StrictPlatformModel):
    revision: int = Field(ge=1, le=MAX_SAFE_INTEGER)
    interval_seconds: PlatformIntervals
    updated_at: datetime

    @property
    def intervals_seconds(self) -> PlatformIntervals:
        return self.interval_seconds


class PlatformAccessUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    expected_revision: int = Field(ge=1, le=MAX_SAFE_INTEGER)
    interval_seconds: PlatformIntervals


class PlatformAccessDiagnostic(StrictPlatformModel):
    """Safe evidence projected when a platform blocks an application request."""

    platform: SearchPlatform
    stage: Literal["search", "detail", "media"]
    outcome: Literal["platform_blocked_or_rate_limited"]
    basis: Literal["http_status", "explicit_platform_evidence", "browser_dom_evidence"]
    status_code: int | None = Field(default=None, ge=100, le=599)
    platform_code: str | None = Field(default=None, min_length=1, max_length=100)
    retry_after_at: datetime | None = None
    manual_challenge_required: bool = False
    observed_at: datetime

    @model_validator(mode="after")
    def normalize_timestamps(self):
        for field in ("retry_after_at", "observed_at"):
            value = getattr(self, field)
            if value is not None and value.tzinfo is None:
                object.__setattr__(self, field, value.replace(tzinfo=UTC))
        return self


def default_platform_intervals() -> PlatformIntervals:
    return PlatformIntervals(
        **{platform: PLATFORM_ACCESS_DEFAULT_SECONDS for platform in SEARCH_PLATFORMS}
    )


def default_platform_access_snapshot(
    *, basis: PlatformAccessBasis = "upgrade_safe_default"
) -> PlatformAccessSnapshot:
    return PlatformAccessSnapshot(
        interval_seconds=default_platform_intervals(), basis=basis
    )
